src/main/java/org/aksw/simba/lemming/metrics/single/AvgVertexDegreeMetric.java

```diff
@@ -1,37 +1,151 @@
+	 * @param graph
+	 *            - input graph.
+	 * @param graphOperation
+	 *            - boolean value indicating graph operation. ("true" for adding an
+	 *            edge and "false" for removing an edge)
+	 * @param previousResult
+	 *            - UpdatableMetricResult object containing the previous computed
+	 *            results.
+	 * @return UpdatableMetricResult object with updated values that can be used in
+	 *         further computations
+	 */
+	@Override
+	public UpdatableMetricResult update(TripleBaseSingleID triple, ColouredGraph graph, boolean graphOperation,
+			UpdatableMetricResult previousResult, VertexDegrees mVertexDegrees) {
+
+		if (getName().equals("stdDevInDegree") || getName().equals("stdDevOutDegree")) {
+			return new SimpleMetricResult(getName(), apply(graph));
+		}
+
+		AvgVertexDegreeMetricResult metricResultTempObj = (AvgVertexDegreeMetricResult) applyUpdatable(graph, previousResult);
+
+		double sum = 0;
+		double numberOfVertices = 1;
+		if (graphOperation) { // If Add an Edge
+
+			if (metricResultTempObj.getSumVertexDeg() == 0.0) {
+				// Computing the Avg Vertex Degree Metric for the first time
+
+				// Get the Array from VertexDegrees class (Note: This can be replaced with
+				// getAllInEdgeDegrees method of Grph package)
+				int[] getmMapVerticesinDegree = mVertexDegrees.getmMapVerticesinDegree();
+				for (int key = 0; key < getmMapVerticesinDegree.length; key++) { // Compute sum in iteration
+					sum += getmMapVerticesinDegree[key];
+				}
+				numberOfVertices = getmMapVerticesinDegree.length;
+
+			} else { // Re-using the previously computed values
+				sum = metricResultTempObj.getSumVertexDeg() + 1;
+				// Get the previous computed sum and add 1 to previous sum since edge is added.
+				numberOfVertices = metricResultTempObj.getNumberOfVertices();
+			}
+
+			// Set values in Temporary object
+			metricResultTempObj.setSumVertexDeg(sum);
+			metricResultTempObj.setNumberOfVertices(numberOfVertices);
+
+		} else { // If Remove an Edge
+
+			if (metricResultTempObj.getSumVertexDeg() == 0.0) {
+				// Computing the Avg Vertex Degree Metric for the first time
+
+				int[] getmMapVerticesinDegree = mVertexDegrees.getmMapVerticesinDegree();
+
+				for (int key = 0; key < getmMapVerticesinDegree.length; key++) { // Compute sum in iteration
+					sum += getmMapVerticesinDegree[key];
+				}
+				numberOfVertices = getmMapVerticesinDegree.length;
+
+			} else { // Re-using the previously computed values
+				sum = metricResultTempObj.getSumVertexDeg() - 1;
+				// Get the previous computed sum and subtract 1 to previous sum since edge is
+				// removed.
+
+				numberOfVertices = metricResultTempObj.getNumberOfVertices();
+
+			}
+
+			metricResultTempObj.setSumVertexDeg(sum);
+			metricResultTempObj.setNumberOfVertices(numberOfVertices);
+		}
+
+		sum = sum / numberOfVertices; // Compute Metric value
+		metricResultTempObj.setResult(sum);
+
+		return metricResultTempObj;
```

Comment:
This can be further optimized in an easy way. Please think about which piece of information you _really_ need when you want to get the average degree, degree standard deviation or one of the minimum or maximum values.

Reply:
For the standard deviation metric, logic was created in StdDevVertexDegree class (update method) to calculate it in an optimized way. These changes are implemented by Pranav and are available in the optimize_stddev_metric branch. Thus, this method will not be called for standard deviation metric once those changes are included. We are checking further to implement it in a better way.